test/release/examples/patterns/readcsv.chpl

```diff
@@ -0,0 +1,286 @@
+    file readers and writers
+      https://chapel-lang.org/docs/main/modules/standard/IO.html#i-o-overview
+      https://chapel-lang.org/docs/primers/fileIO.html
+
+    lists
+      https://chapel-lang.org/docs/primers/listOps.html
+
+    maps
+      https://chapel-lang.org/docs/main/modules/standard/Map.html
+ */
+use IO;
+use List;
+use Map;
+
+config const inFileName = "metadata-20lines.csv";
+config const approach1FileName = "readcsv-out1.csv";
+config const approach2FileName = "readcsv-out2.csv";
+config const debug = true;
+
+// Open up a file to work with.
+var f = open(inFileName, iomode.r);
+
+// Now read the in the csv data. 
+// Approach 1: formatted input into a list of maps
+{
+  writeln();
+  writeln();
+  writeln("Approach 1: formatted input into a list of maps");
+
+  var reader = f.reader();
+
+  // Read the first line to get the column names.
+  // Assuming the following format, where the number of columns (ncol) is not 
+  // known ahead of time:
+  //    col_name_1,col_name_2,...,col_name_ncol\n
+  var line : string;
+
+  if (!reader.readline(line)) then
+    writeln("ERROR: ", inFileName, " appears to be empty");
+
+  var colNames : list(string);
+  colNames = createListOfColNames(line);
+
+  if debug {
+    writeln("\ncolNames: ", colNames);
+  }
+
+  // Create a list of maps with one map per row of data.
+  var dataRows : list(map(string,string));
+  // Read individual lines from the file until the end of the file.
+  while (reader.readline(line)) {
+    // Create a new map per line.
+    var aRowMap = new map(string, string);
+    var start = 0;      // index into the line string
+    var colIdx = 0;     // the first column is indexed as zero
+    // Process each line by finding the next field value from the start index to the next comma.
+    while (start<line.size && colIdx<colNames.size) {
+      var (nextVal,nextCommaIdx) = nextField(line,start);
+      aRowMap.add(colNames[colIdx],nextVal);
+      start = nextCommaIdx + 1;         // The next column of data will start after the comma.
+      colIdx = colIdx + 1;              // Increment the column index.
+    }
+    dataRows.append(aRowMap);
+  }
+  reader.close();
+
+  // Print out the last data row to illustrate what one of the row maps contains.
+  if debug then writeln("\ndataRows[dataRows.size-1] = ", dataRows[dataRows.size-1]);
+ 
+  //-----------------------------------------------------------------------
+  // To see how the data is accessed and for testing, let's generate a 
+  // temporary csv file from the stored data.  It should be the same as the 
+  // input csv file.
+  if debug {
+    var outfile = open(approach1FileName, iomode.cw);
+    var writer = outfile.writer();
+
+    // First write to the output file the column names separated by commas.
+    for colIdx in 0..colNames.size-2 {
+      writer.write(colNames[colIdx],",");
+    }
+    // The last column name should not be followed by comma.
+    writer.writeln(colNames[colNames.size-1]);
```

Comment:
All of this printing out with commas in between could be done with:

```chapel
writer.writeln(",".join(colNames.these()));
```

It is still a bit unfortunate, though. Because I'm not sure if we want users to use `these` explicitly like this. I thought we were planning to do https://github.com/chapel-lang/chapel/issues/17252 for the release, but we couldn't get to it in time apparently.

Reply:
Yeah I am going to leave this one as is so as to keep the conceptual burden low.